### PR TITLE
Allow override of ad template on property

### PR DESCRIPTION
**Pull request by hopsoft**
##### Description

> Enables both the administrators and publishers to be able to enhance the display ad the publisher's websites without having to make changes to the code.

##### Refers/Fixes

Fixes: #228

### CODE REVIEW

**Comment by hopsoft** [comment]
Basic rendering in the admin.

codefund 2018-07-25 22-58-48

**Comment by hopsoft** [comment]
Note that I've moved away from drop downs to text fields for a simpler solution.

**Comment by coderberry** [comment]
@hopsoft why move away from dropdowns? It makes more sense to use those. That way the developer doesn't need to guess which ones exist

**Comment by hopsoft** [comment]
@coderberry I moved back to a dropdown but have omitted the `theme` dropdown for now. I figure we can address `theme` overrides in another PR. 

The basic work is done, but I could use some pair time with either @coderberry or @StareIntoTheBeard to review the code & get a brain dump on how you do local testing for ad rendering. Also, I need to get some unit tests wrapped around this work.

File: lib/code_fund_web/controllers/api/ad_serve_controller.ex [hunk 1]
```diff
@@ -30,6 +30,15 @@ defmodule CodeFundWeb.API.AdServeController do
     end
   end
 
+  defp get_template_slug(property_id, requested_template_slug) do
```
Comment (StareIntoTheBeard) [inline lib/code_fund_web/controllers/api/ad_serve_controller.ex]: Functions which look up objects from the database don't belong in the controller. If it were me, I'd move something like this as a public function in Templates, join to properties where property.id == property_id and I'd rewrite the case as a pattern matched function clause def that the slug.

```
defmodule Templates do
  def for_property_id(property_id, requested_template_slug) do
     from(t in Template, join: p in Property on: p.template_id == t.id,
     where: p.id == ^property_id)
     |> Repo.one
     |> return_template(requested_template_slug)  
   end
  defp return_template(%Template{slug}}, requested_template_slug), when not(is_nil(slug)), do: slug
  defp return_template(_,_), do: requested_template_slug
end

File: lib/code_fund_web/controllers/api/ad_serve_controller.ex [hunk 2]
```diff
@@ -5,9 +5,9 @@ defmodule CodeFundWeb.API.AdServeController do
   alias CodeFund.Schema.{Campaign, Impression, Property, Theme, Template}
 
   def embed(conn, %{"property_id" => property_id} = params) do
```
Comment (StareIntoTheBeard) [inline lib/code_fund_web/controllers/api/ad_serve_controller.ex]: @hopsoft why not match "template" out of params here too?

File: lib/code_fund_web/views/property_view.ex [hunk 3]
```diff
@@ -10,11 +10,13 @@ defmodule CodeFundWeb.PropertyView do
     "//www.google.com/s2/favicons?domain=#{domain}"
   end
 
-  def script_embed_code(conn, %CodeFund.Schema.Property{} = property) do
-    url = "https://#{conn.host}/scripts/#{property.id}/embed.js"
+  def script_embed_url(conn, %CodeFund.Schema.Property{} = property) do
```
Comment (StareIntoTheBeard) [inline lib/code_fund_web/views/property_view.ex]: @hopsoft can we pull the property id and possibly conn.host out of the struct in the function def so we don't have to match later?

``` def script_embed_url(%Plug.Conn{host: host}, %CodeFund.Schema.Property{id: id} ) ```

File: lib/code_fund_web/controllers/api/ad_serve_controller.ex (hunk 1 above)
Comment (hopsoft) [reply lib/code_fund_web/controllers/api/ad_serve_controller.ex]: This is great feedback. Thanks for pointing me in the right direction.

File: lib/code_fund_web/controllers/api/ad_serve_controller.ex (hunk 2 above)
Comment (hopsoft) [reply lib/code_fund_web/controllers/api/ad_serve_controller.ex]: Attempted this but get test failures similar to this when matching `template`.

```
** (Phoenix.ActionClauseError) could not find a matching CodeFundWeb.API.AdServeController.embed clause
     to process request. This typically happens when there is a
     parameter mismatch but may also happen when any of the other
     action arguments do not match. The request parameters are:

       %{"property_id" => "bae186f2-2489-43c9-8962-7f09cae083fc"}
```

File: lib/code_fund_web/controllers/api/ad_serve_controller.ex (hunk 1 above)
Comment (StareIntoTheBeard) [reply lib/code_fund_web/controllers/api/ad_serve_controller.ex]: You bet

**Comment by hopsoft** [comment]
Refactors based on feedback are up. Ready for re-review.

**Pull request merged**